### PR TITLE
Shape change bitcast for more dtypes

### PR DESCRIPTION
**Pull request by samm393**
Generalised version of #6043

### CODE REVIEW

**Comment by geohot** [comment]
Actually, this is better than I first thought. Sorry, I've been frustrated by all the terrible PRs in bitcast. I'll review it

File: test/test_dtype.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ class TestDType(unittest.TestCase):
   @classmethod
   def setUpClass(cls):
     if not cls.DTYPE or not is_dtype_supported(cls.DTYPE): raise unittest.SkipTest("dtype not supported")
-    cls.DATA = rand_for_dtype(cls.DTYPE, 10)
+    cls.DATA = rand_for_dtype(cls.DTYPE, 8)
```
Comment (geohot) [inline test/test_dtype.py:66]: This is a global change to all dtype tests, is that what you intended?

**Comment by geohot** [comment]
There's a lack of tests for multidimensional tensors being bitcasted. Also, need to get the new remu instructions implemented (that's what's causing AMD to fail)

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -3062,7 +3062,14 @@ def bitcast(self, dtype:DTypeLike) -> Tensor:
...
     """
     if self.requires_grad: raise RuntimeError("can't backprop through bitcast")
-    return F.Cast.apply(self, dtype=dt, bitcast=True) if self.dtype != (dt:=to_dtype(dtype)) else self
+    dt = to_dtype(dtype)
+    if (not isinstance(self.device, str) or not self.device.startswith("DISK")) and (ns:=dt.itemsize) != (os:=self.dtype.itemsize):
+      if not (self.shape[-1]*os) % ns == 0: raise RuntimeError("unsupported size in bitcast")
+      new_uint, old_uint = to_dtype(f"uint{8*ns}"), to_dtype(f"uint{8*os}")
+      tmp = self.bitcast(old_uint)
+      if ns > os: return functools.reduce(Tensor.__add__, (tmp[...,i::ns//os].cast(new_uint) << 8*i*os for i in range(ns//os))).bitcast(dtype)
+      return Tensor.stack(*(tmp >> 8*i*ns for i in range(os//ns)), dim=-1).flatten(-2).cast(new_uint).bitcast(dtype)
```
Comment (geohot) [inline tinygrad/tensor.py:3071]: Is this cast guaranteed to correctly truncate, or do you need & 0xFF...

**Comment by geohot** [comment]
@Qazalin seeing some remu failures here. How close are we to 100% completeness of the ISA?

**Comment by Qazalin** [comment]
released new instruction, 1 line https://github.com/Qazalin/remu/commit/06daa42dc26d510266afe2714a41318fb385eab4.
very close, remu has infra for vector, scalar, lds, gds and scratch formats.

File: test/test_dtype.py (hunk 1 above)
Comment (samm393) [reply test/test_dtype.py]: Reverted, have used DATA[:8] instead for the bitcast test

File: tinygrad/tensor.py (hunk 2 above)
Comment (geohot) [reply tinygrad/tensor.py]: Yea, it's better without if it you are confident it's correct.

**Comment by geohot** [comment]
Glad this is finally over. Good work!

**Comment by geohot** [comment]
If you want half the bounty ($125) it's yours, can pay with PayPal or USDC, george@tinygrad.org to claim

**Pull request merged**